editor/src/clj/editor/particlefx.clj

```diff
@@ -386,18 +398,30 @@
 (defn- convert-blend-mode [blend-mode-index]
   (protobuf/pb-enum->val (.getValueDescriptor (Particle$BlendMode/valueOf ^int blend-mode-index))))
 
-(defn- render-emitters-sim [^GL2 gl render-args renderables rcount]
+(defn- vset [v i value]
+  (let [c (count v)
+        v (if (<= c i) (into v (repeat (- i c) nil)) v)]
+    (assoc v i value)))
+
+(defn- render-emitters-sim [^GL2 gl render-args renderables _rcount]
   (doseq [renderable renderables]
-    (let [{:keys [emitter-sim-data emitter-index color]} (:user-data renderable)
+    (let [user-data (:user-data renderable)
+          {:keys [emitter-sim-data emitter-index color max-particle-count]} user-data
+          shader (:shader emitter-sim-data)
+          shader-bound-attributes (graphics/get-shader-bound-attributes gl shader (:material-attribute-infos user-data) [:position :texcoord0 :page-index :color])
+          vertex-description (graphics/make-vertex-description shader-bound-attributes)
+          vertex-attribute-bytes (:vertex-attribute-bytes user-data)
           pfx-sim-request-id (some-> renderable :updatable :node-id)]
-      (when-let [pfx-sim (when (and emitter-sim-data pfx-sim-request-id)
-                           (some-> (:pfx-sim (scene-cache/lookup-object ::pfx-sim pfx-sim-request-id nil)) deref))]
-        (plib/gen-emitter-vertex-data pfx-sim emitter-index color)
-        (let [context (:context pfx-sim)
-              vbuf (plib/get-emitter-vertex-data pfx-sim emitter-index)]
+      (when-let [pfx-sim-atom (when (and emitter-sim-data pfx-sim-request-id)
+                                (:pfx-sim (scene-cache/lookup-object ::pfx-sim pfx-sim-request-id nil)))]
+        (let [pfx-sim @pfx-sim-atom
+              raw-vbuf (plib/gen-emitter-vertex-data pfx-sim emitter-index color max-particle-count vertex-description shader-bound-attributes vertex-attribute-bytes)
+              context (:context pfx-sim)
+              vbuf (vtx/wrap-vertex-buffer vertex-description :static raw-vbuf)
+              all-raw-vbufs (vset (:raw-vbufs pfx-sim) emitter-index raw-vbuf)]
+          (swap! pfx-sim-atom assoc :raw-vbufs all-raw-vbufs)
```

Comment:
Yeah, it makes things more complicated for sure... But hm, I'll add a note about this for now and it will have to be a headache for the future I guess 🤷